lib/with_win32.dart

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2020, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:ffi';
+import 'package:ffi/ffi.dart';
+import 'win32.dart' as win32;
+
+export 'win32.dart'
+    hide
+        CreateWindowEx,
+        DispatchMessage,
+        DrawText,
+        GetClientRect,
+        GetMessage,
+        RegisterClass,
+        TranslateMessage;
+
+T withMSG<T>(T Function(win32.MSG) func) {
+  final msg = win32.MSG.allocate();
+  T t;
+  try {
+    t = func(msg);
+  } finally {
+    free(msg.addressOf);
+  }
+  return t;
+}
+
+T withPAINTSTRUCT<T>(T Function(win32.PAINTSTRUCT) func) {
+  final ps = win32.PAINTSTRUCT.allocate();
+  T t;
+  try {
+    t = func(ps);
+  } finally {
+    free(ps.addressOf);
+  }
+  return t;
+}
+
+T withRECT<T>(T Function(win32.RECT) func) {
+  final rect = win32.RECT.allocate();
+  T t;
+  try {
+    t = func(rect);
+  } finally {
+    free(rect.addressOf);
+  }
+  return t;
+}
+
+T withTEXT<T>(String text, T Function(Pointer<Utf16>) func) {
+  final utf16Text = win32.TEXT(text);
+  T t;
+  try {
+    t = func(utf16Text);
+  } finally {
+    free(utf16Text);
+  }
+  return t;
+}
+
+T withWNDCLASS<T>(T Function(win32.WNDCLASS) func) {
+  final wc = win32.WNDCLASS.allocate();
+  T t;
+  try {
+    t = func(wc);
+  } finally {
+    free(wc.addressOf);
+  }
+  return t;
+}
+
+T BeginPaint<T>(int hWnd, T Function(int hdc) func) => withPAINTSTRUCT((ps) {
+      final hdc = win32.BeginPaint(hWnd, ps.addressOf);
+      T t;
+      try {
+        t = func(hdc);
+      } finally {
+        win32.EndPaint(hWnd, ps.addressOf);
+      }
+      return t;
+    });
+
+int CreateWindowEx(
+        int dwExStyle,
+        String className,
```

Comment:
My understanding is that this `LPCTSTR` must be the same one as created through `RegisterClass`.  I haven't tried allocating the same string with a different pointer address, but normally Win32 apps hold a reference to this string and pass it through.

Reply:
That would be weird. I understand why C/C++ developers create a string once and hold onto it for the lifetime of the code, their language basically teaches people to be afraid of `malloc` and `free`. Most of the time it'd be a static string anyway, and thus packed into the PE, and not generated at run time.

With that said, I'd find it odd for the `CreateWindowEx` to only do pointer comparison, as opposed to string comparison. Most libc string comparison implementations I've seen have a pointer comparison fast path anyway, so it's a minimal optimisation that would break the explicit API contract. Is it documented as required to be pointer identical somewhere? The [CreateWindowEx docs](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-createwindowexa) make no mention of pointer equality as a test.

This code appears to work as is, and IIUC this is also how the Rust team are approaching Win32 interop.